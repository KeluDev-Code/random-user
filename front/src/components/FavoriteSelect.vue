<template>
  <q-dialog
    ref="dialogRef"
    @hide="onDialogHide"
  >
    <q-card class="q-dialog-plugin">
      <q-card-section>
        <div class="text-h6">
          {{ $t('Acctions.addFavorite') }}
        </div>
      </q-card-section>
      <q-card-section>
        <div class="q-px-sm">
          <div class="text-subtitle1">
            {{ $t('Labels.addToList') }}
          </div>
          <div class="q-pa-sm">
            <q-select
              v-model="favoriteSelected"
              :disable="!!nickName"
              :label="$t('Fields.favorites')"
              :options="favorites"
              option-label="nickName"
              dense
              outlined
              clearable
              class="bg-white"
            />
          </div>

          <q-separator class="q-my-sm" />

          <div class="text-subtitle1">
            {{ $t('Labels.createList') }}
          </div>
          <div class="q-pa-sm">
            <q-input
              v-model="nickName"
              :label="$t('Fields.favorites')"
              :disable="!!favoriteSelected"
              dense
              outlined
              clearable
              class="bg-white"
            />
          </div>
        </div>
      </q-card-section>

      <q-card-actions align="right">
        <q-btn
          color="negative"
          flat
          :label="$t('Labels.cancel')"
          @click="onDialogCancel"
        />
        <q-btn
          color="primary"
          :disable="!favoriteSelected && !nickName"
          :label="$t('Labels.ok')"
          @click="onOKClick"
        />
      </q-card-actions>
    </q-card>
  </q-dialog>
</template>

<script lang="ts">
import { useDialogPluginComponent } from 'quasar';
import { onMounted, ref } from 'vue';

// eslint-disable-next-line import/no-cycle
import useFavorite from '@/services/FavoriteService';

export default {
  name: 'FavoriteSelect',
  emits: [
    ...useDialogPluginComponent.emits,
  ],

  setup() {
    const usedFavorite = useFavorite();
    const nickName = ref('');

    const {
      dialogRef, onDialogHide, onDialogOK, onDialogCancel,
    } = useDialogPluginComponent();
    // dialogRef      - Vue ref to be applied to QDialog
    // onDialogHide   - Function to be used as handler for @hide on QDialog
    // onDialogOK     - Function to call to settle dialog with "ok" outcome
    // onDialogCancel - Function to call to settle dialog with "cancel" outcome
    const onOKClick = () => {
      // on OK, it is REQUIRED to call onDialogOK
      onDialogOK(usedFavorite.favoriteSelected.value?.nickName || nickName);
    };

    onMounted(() => {
      usedFavorite.loadFavorites();
    });

    return {
      // REQUIRED; Need to inject these (from useDialogPluginComponent() call)
      dialogRef,
      onDialogHide,
      // other methods that we used in our vue html template;
      // these are part of our example (so not required)
      onOKClick,

      onDialogCancel,

      favoriteSelected: usedFavorite.favoriteSelected,
      favorites: usedFavorite.favorites,
      nickName,
    };
  },
};
</script>
