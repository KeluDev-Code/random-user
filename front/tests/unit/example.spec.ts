/* eslint-disable @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any, @typescript-eslint/ban-ts-comment */
import {
  mount, shallowMount, VueWrapper, config,
} from '@vue/test-utils';
import { useI18n } from 'vue-i18n';
import fetchMock from 'jest-fetch-mock';
import {
  QCardSection, QSelect, QSeparator, QInput, QBtn, QCardActions, QCard, QDialog,
} from 'quasar';

import FavoriteSelect from '@/components/FavoriteSelect.vue';

// @ts-ignore
//
const QComponents = {
  QCardSection, QSelect, QSeparator, QInput, QBtn, QCardActions, QCard, QDialog,
};

// establecemos globalmente un mock de $t
// config.global.mocks.$t = (key:string) => key;
// establecemos globalmente los components
// config.global.components = QComponents;

// mock sobre la libreria 'vue-i18n' donde solo se hace el mock a 'useI18n'
jest.mock(
  'vue-i18n',
  () => {
    const mUseI18n = { t: (key:string) => key };
    return {
      ...(jest.requireActual('vue-i18n')),
      useI18n: jest.fn(() => mUseI18n),
    };
  },
  { virtual: true },
);

describe('FavoriteSelect.vue', () => {
  let wrapper : VueWrapper<any>;

  beforeEach(() => {
    fetchMock.doMock();

    wrapper = shallowMount(FavoriteSelect, {
      global: {
        // establecemos los components del fichero
        components: QComponents,
      },
    });
  });

  test('debe de mostrarse correctamente', () => {
    // const { html } = wrapper;
    expect(wrapper.element).toMatchSnapshot();
  });

  test('renders props.msg when passed', () => {
    fetchMock.mockResponseOnce(JSON.stringify({ data: '12345' }));

    const DOM = wrapper.find('.q-btn.disable');
    expect(DOM.exists()).toBe(false);
  });
});
