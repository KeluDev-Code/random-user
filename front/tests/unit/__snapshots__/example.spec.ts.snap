// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteSelect.vue HelloWorld debe de mostrarse correctamente 1`] = `
<div class="hello">
  <h1></h1>
  <p> For a guide and recipes on how to configure / customize this project,<br> check out the <a href="https://cli.vuejs.org" target="_blank" rel="noopener">vue-cli documentation</a>. </p>
  <h3>Installed CLI Plugins</h3>
  <ul>
    <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-babel" target="_blank" rel="noopener">babel</a></li>
    <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-pwa" target="_blank" rel="noopener">pwa</a></li>
    <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-router" target="_blank" rel="noopener">router</a></li>
    <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-eslint" target="_blank" rel="noopener">eslint</a></li>
    <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-unit-mocha" target="_blank" rel="noopener">unit-mocha</a></li>
    <li><a href="https://github.com/vuejs/vue-cli/tree/dev/packages/%40vue/cli-plugin-typescript" target="_blank" rel="noopener">typescript</a></li>
  </ul>
  <h3>Essential Links</h3>
  <ul>
    <li><a href="https://vuejs.org" target="_blank" rel="noopener">Core Docs</a></li>
    <li><a href="https://forum.vuejs.org" target="_blank" rel="noopener">Forum</a></li>
    <li><a href="https://chat.vuejs.org" target="_blank" rel="noopener">Community Chat</a></li>
    <li><a href="https://twitter.com/vuejs" target="_blank" rel="noopener">Twitter</a></li>
    <li><a href="https://news.vuejs.org" target="_blank" rel="noopener">News</a></li>
  </ul>
  <h3>Ecosystem</h3>
  <ul>
    <li><a href="https://router.vuejs.org" target="_blank" rel="noopener">vue-router</a></li>
    <li><a href="https://vuex.vuejs.org" target="_blank" rel="noopener">vuex</a></li>
    <li><a href="https://github.com/vuejs/vue-devtools#vue-devtools" target="_blank" rel="noopener">vue-devtools</a></li>
    <li><a href="https://vue-loader.vuejs.org" target="_blank" rel="noopener">vue-loader</a></li>
    <li><a href="https://github.com/vuejs/awesome-vue" target="_blank" rel="noopener">awesome-vue</a></li>
  </ul>
</div>
`;

exports[`FavoriteSelect.vue debe de mostrarse correctamente 1`] = `
<q-dialog>
  <q-card class="q-dialog-plugin">
    <q-card-section>
      <div class="text-h6">Acctions.addFavorite</div>
    </q-card-section>
    <q-card-section>
      <div class="q-px-sm">
        <div class="text-subtitle1">Labels.addToList</div>
        <div class="q-pa-sm">
          <q-select disable="true" label="Fields.favorites" options="" option-label="nickName" dense="" outlined="" clearable="" class="bg-white"></q-select>
        </div>
        <q-separator class="q-my-sm"></q-separator>
        <div class="text-subtitle1">Labels.createList</div>
        <div class="q-pa-sm">
          <q-input modelvalue="" label="Fields.favorites" disable="false" dense="" outlined="" clearable="" class="bg-white"></q-input>
        </div>
      </div>
    </q-card-section>
    <q-card-actions align="right">
      <q-btn color="negative" flat="" label="Labels.cancel"></q-btn>
      <q-btn color="primary" disable="true" label="Labels.ok"></q-btn>
    </q-card-actions>
  </q-card>
</q-dialog>
`;
